File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/engine/preparation/PreparedPlaybackQueueFeederBuilder.java
package com.lasthopesoftware.bluewater.client.playback.engine.preparation;

import android.os.Handler;

import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.RenderersFactory;
import com.lasthopesoftware.bluewater.client.library.items.media.files.uri.BestMatchUriProvider;
import com.lasthopesoftware.bluewater.client.library.repository.Library;
import com.lasthopesoftware.bluewater.client.playback.engine.exoplayer.queued.QueueMediaSources;
import com.lasthopesoftware.bluewater.client.playback.engine.selection.LookupSelectedPlaybackEngineType;
import com.lasthopesoftware.bluewater.client.playback.file.exoplayer.preparation.ExoPlayerPlayableFilePreparationSourceProvider;
import com.lasthopesoftware.bluewater.client.playback.file.exoplayer.preparation.SingleExoPlayerSourcePreparationSourceProvider;
import com.lasthopesoftware.bluewater.client.playback.file.exoplayer.preparation.mediasource.ExtractorMediaSourceFactoryProvider;
import com.namehillsoftware.handoff.promises.Promise;

public class PreparedPlaybackQueueFeederBuilder implements BuildPreparedPlaybackQueueFeeder {

	private final LookupSelectedPlaybackEngineType selectedPlaybackEngineTypeLookup;
	private final Handler handler;
	private final BestMatchUriProvider bestMatchUriProvider;
	private final ExtractorMediaSourceFactoryProvider mediaSourceFactoryProvider;
	private final ExoPlayer exoPlayer;
	private final QueueMediaSources mediaSourcesQueue;
	private final RenderersFactory renderersFactory;

	public PreparedPlaybackQueueFeederBuilder(
		LookupSelectedPlaybackEngineType selectedPlaybackEngineTypeLookup,
		Handler handler,
		BestMatchUriProvider bestMatchUriProvider,
		ExtractorMediaSourceFactoryProvider mediaSourceFactoryProvider,
		ExoPlayer exoPlayer,
		QueueMediaSources mediaSourcesQueue,
		RenderersFactory renderersFactory) {

		this.selectedPlaybackEngineTypeLookup = selectedPlaybackEngineTypeLookup;
		this.handler = handler;
		this.bestMatchUriProvider = bestMatchUriProvider;
		this.mediaSourceFactoryProvider = mediaSourceFactoryProvider;
		this.exoPlayer = exoPlayer;
		this.mediaSourcesQueue = mediaSourcesQueue;
		this.renderersFactory = renderersFactory;
	}

	@Override
	public Promise<IPlayableFilePreparationSourceProvider> build(Library library) {
		return this.selectedPlaybackEngineTypeLookup.promiseSelectedPlaybackEngineType()
			.then(playbackEngineType -> {
				switch (playbackEngineType) {
					case ExoPlayer:
						return new ExoPlayerPlayableFilePreparationSourceProvider(
							handler,
							bestMatchUriProvider,
							mediaSourceFactoryProvider,
							renderersFactory);
					case SingleExoPlayer:
						return new SingleExoPlayerSourcePreparationSourceProvider(
							handler,
							bestMatchUriProvider,
							mediaSourceFactoryProvider,
							exoPlayer,
							mediaSourcesQueue,
							renderersFactory);
					default:
						return null;
				}
			});
	}
}

File: playback/src/main/java/com/namehillsoftware/projectblue/playback/engine/preparation/PreparedPlaybackQueueFeederBuilder.kt
package com.lasthopesoftware.bluewater.client.playback.engine.preparation

import android.content.Context
import android.os.Handler
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.uri.BestMatchUriProvider
import com.lasthopesoftware.bluewater.client.browsing.library.repository.Library
import com.lasthopesoftware.bluewater.client.playback.file.exoplayer.preparation.ExoPlayerPlayableFilePreparationSourceProvider
import com.lasthopesoftware.bluewater.client.playback.file.exoplayer.preparation.mediasource.SpawnMediaSources

class PreparedPlaybackQueueFeederBuilder(
        private val context: Context,
        private val handler: Handler,
        private val mediaSourceProvider: com.lasthopesoftware.bluewater.client.playback.file.exoplayer.preparation.mediasource.SpawnMediaSources,
        private val bestMatchUriProvider: BestMatchUriProvider) : BuildPreparedPlaybackQueueFeeder {

	override fun build(library: Library): IPlayableFilePreparationSourceProvider =
            com.lasthopesoftware.bluewater.client.playback.file.exoplayer.preparation.ExoPlayerPlayableFilePreparationSourceProvider(
                    context,
                    handler,
                    mediaSourceProvider,
                    bestMatchUriProvider)
}
